chain/sync_manager.go

```diff
@@ -0,0 +1,369 @@
+
+func (stb *syncTargetBucket) add(ts *types.TipSet) {
+	stb.count++
+
+	for _, t := range stb.tips {
+		if t.Equals(ts) {
+			return
+		}
+	}
+
+	stb.tips = append(stb.tips, ts)
+}
+
+func (stb *syncTargetBucket) heaviestTipSet() *types.TipSet {
+	if stb == nil {
+		return nil
+	}
+
+	var best *types.TipSet
+	for _, ts := range stb.tips {
+		if best == nil || ts.ParentWeight().GreaterThan(best.ParentWeight()) {
+			best = ts
+		}
+	}
+	return best
+}
+
+func (sm *SyncManager) selectSyncTarget() (*types.TipSet, error) {
+	var buckets syncBucketSet
+
+	var peerHeads []*types.TipSet
+	for _, ts := range sm.peerHeads {
+		peerHeads = append(peerHeads, ts)
+	}
+	sort.Slice(peerHeads, func(i, j int) bool {
+		return peerHeads[i].Height() < peerHeads[j].Height()
+	})
+
+	for _, ts := range peerHeads {
+		buckets.Insert(ts)
+	}
+
+	if len(buckets.buckets) > 1 {
+		log.Warning("caution, multiple distinct chains seen during head selections")
+		// TODO: we *could* refuse to sync here without user intervention.
+		// For now, just select the best cluster
+	}
+
+	return buckets.Heaviest(), nil
+}
+
+func (sm *SyncManager) syncScheduler() {
+
+	for {
+		select {
+		case ts, ok := <-sm.incomingTipSets:
+			if !ok {
+				log.Info("shutting down sync scheduler")
+				return
+			}
+
+			sm.scheduleIncoming(ts)
+		case res := <-sm.syncResults:
+			sm.scheduleProcessResult(res)
+		case sm.workerChan <- sm.nextSyncTarget.heaviestTipSet():
+			sm.scheduleWorkSent()
+		case <-sm.stop:
+			log.Info("sync scheduler shutting down")
+			return
+		}
+	}
+}
+
+func (sm *SyncManager) scheduleIncoming(ts *types.TipSet) {
+	var relatedToActiveSync bool
+	for _, acts := range sm.activeSyncs {
+		if ts.Equals(acts) {
+			break
+		}
+
+		if types.CidArrsEqual(ts.Parents(), acts.Cids()) {
+			// sync this next, after that sync process finishes
+			relatedToActiveSync = true
+		}
+	}
+
+	// if this is related to an active sync process, immediately bucket it
+	// we don't want to start a parallel sync process that duplicates work
+	if relatedToActiveSync {
+		sm.activeSyncTips.Insert(ts)
+		return
+	}
+
+	if sm.nextSyncTarget != nil && sm.nextSyncTarget.sameChainAs(ts) {
+		sm.nextSyncTarget.add(ts)
+	} else {
+		sm.syncQueue.Insert(ts)
+
+		if sm.nextSyncTarget == nil {
+			sm.nextSyncTarget = sm.syncQueue.Pop()
+			sm.workerChan = sm.syncTargets
+		}
+	}
+}
+
+func (sm *SyncManager) scheduleProcessResult(res *syncResult) {
+	delete(sm.activeSyncs, res.ts.Key())
+	relbucket := sm.activeSyncTips.PopRelated(res.ts)
+	if relbucket != nil {
+		if res.success {
+			if sm.nextSyncTarget == nil {
+				sm.nextSyncTarget = relbucket
+				sm.workerChan = sm.syncTargets
+			} else {
+				sm.syncQueue.buckets = append(sm.syncQueue.buckets, relbucket)
+			}
+		} else {
+			// TODO: this is the case where we try to sync a chain, and
+			// fail, and we have more blocks on top of that chain that
+			// have come in since.  The question is, should we try to
+			// sync these? or just drop them?
```

Comment:
If we can retry later I'd say dropping them is fine

Reply:
Yeah, we can definitely retry later. I think I agree, dropping is fine